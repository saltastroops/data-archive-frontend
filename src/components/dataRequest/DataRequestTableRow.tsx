import * as React from "react";
import { IDataFile, IDataRequestPart } from "./DataRequestsForm";

/**
 * Props of the data request table row component.
 *
 * Properties:
 * -----------
 * dataRequestPart:
 *     Part of the data request displayed in the row.
 */

interface IDataRequestTableRowProps {
  dataRequestPart: IDataRequestPart;
}

/**
 * A row in the data request table.
 */
class DataRequestTableRow extends React.Component<IDataRequestTableRowProps> {
  render() {
    const { dataFiles, status, uri } = this.props.dataRequestPart;
    let downloadButton = false;
    let reRequestButton = false;
    let statusText = "";

    // Converting the status for meaningful display
    switch (status) {
      case "SUCCESSFUL":
        statusText = "Available";
        downloadButton = true;
        break;
      case "FAILED":
        statusText = "Failed";
        reRequestButton = true;
        break;
      case "PENDING":
        statusText = "Pending";
        break;
      case "EXPIRED":
        statusText = "Expired";
        reRequestButton = true;
        break;
      default:
        statusText = "Unknown";
        break;
    }

    return (
      <tr>
        <td>{dataFiles[0].observation.name}</td>
        <td>
          <p>{statusText}</p>
          <p>
            {downloadButton && (
              <a
                className="button download is-small is-success is-rounded"
                href={uri}
              >
                Download
              </a>
            )}
            {reRequestButton && (
              <button className="button re-request is-small is-danger is-rounded">
                Re-request
              </button>
            )}
          </p>
        </td>
        <td>
          <ul>
            {dataFiles.map((file: IDataFile) => {
              return <li key={file.id}>{file.name}</li>;
            })}
          </ul>
        </td>
      </tr>
    );
  }
}

export default DataRequestTableRow;
