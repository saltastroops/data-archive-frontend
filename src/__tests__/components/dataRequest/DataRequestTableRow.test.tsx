import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import {
  DataRequestStatus,
  IDataRequest
} from "../../../components/dataRequest/DataRequestsForm";
import DataRequestTableRow from "../../../components/dataRequest/DataRequestTableRow";

const dummyDataRequestPartWrapper = (status: DataRequestStatus) => {
  const dataRequestPart = {
    dataFiles: [
      {
        id: "file-1",
        name: "File 1",
        observation: {
          id: "obs-1",
          name: "Observation 1"
        }
      },
      {
        id: "file-2",
        name: "File 2",
        observation: {
          id: "obs-2",
          name: "Observation 2"
        }
      }
    ],
    id: "part-1",
    status,
    uri: "http://demo/data-request/part-1"
  };
  return mount(
    <table>
      <tbody>
        <DataRequestTableRow dataRequestPart={dataRequestPart} />
      </tbody>
    </table>
  );
};

describe("DataRequestTableRow", () => {
  it("renders correctly", () => {
    let wrapper = dummyDataRequestPartWrapper("SUCCESSFUL");
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper = dummyDataRequestPartWrapper("PENDING");
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper = dummyDataRequestPartWrapper("FAILED");
    expect(toJson(wrapper)).toMatchSnapshot();

    wrapper = dummyDataRequestPartWrapper("EXPIRED");
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("includes and excludes download and re-request links as required", async () => {
    let wrapper = dummyDataRequestPartWrapper("SUCCESSFUL");
    expect(wrapper.find(".download").length).toBe(1);
    expect(wrapper.find(".re-request").length).toBe(0);

    wrapper = dummyDataRequestPartWrapper("PENDING");
    expect(wrapper.find(".download").length).toBe(0);
    expect(wrapper.find(".re-request").length).toBe(0);

    wrapper = dummyDataRequestPartWrapper("FAILED");
    expect(wrapper.find(".download").length).toBe(0);
    expect(wrapper.find(".re-request").length).toBe(1);

    wrapper = dummyDataRequestPartWrapper("EXPIRED");
    expect(wrapper.find(".download").length).toBe(0);
    expect(wrapper.find(".re-request").length).toBe(1);
  });
});
