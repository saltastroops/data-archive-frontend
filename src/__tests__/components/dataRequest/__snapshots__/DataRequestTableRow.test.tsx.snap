// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRequestTableRow renders correctly 1`] = `
<table>
  <tbody>
    <DataRequestTableRow
      dataRequestPart={
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-1",
              "name": "File 1",
              "observation": Object {
                "id": "obs-1",
                "name": "Observation 1",
              },
            },
            Object {
              "id": "file-2",
              "name": "File 2",
              "observation": Object {
                "id": "obs-2",
                "name": "Observation 2",
              },
            },
          ],
          "id": "part-1",
          "status": "SUCCESSFUL",
          "uri": "http://demo/data-request/part-1",
        }
      }
    >
      <tr>
        <td>
          Observation 1
        </td>
        <td>
          <p>
            Available
          </p>
          <p>
            <a
              className="button download is-small is-success is-rounded"
              href="http://demo/data-request/part-1"
            >
              Download
            </a>
          </p>
        </td>
        <td>
          <ul>
            <li
              key="file-1"
            >
              File 1
            </li>
            <li
              key="file-2"
            >
              File 2
            </li>
          </ul>
        </td>
      </tr>
    </DataRequestTableRow>
  </tbody>
</table>
`;

exports[`DataRequestTableRow renders correctly 2`] = `
<table>
  <tbody>
    <DataRequestTableRow
      dataRequestPart={
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-1",
              "name": "File 1",
              "observation": Object {
                "id": "obs-1",
                "name": "Observation 1",
              },
            },
            Object {
              "id": "file-2",
              "name": "File 2",
              "observation": Object {
                "id": "obs-2",
                "name": "Observation 2",
              },
            },
          ],
          "id": "part-1",
          "status": "PENDING",
          "uri": "http://demo/data-request/part-1",
        }
      }
    >
      <tr>
        <td>
          Observation 1
        </td>
        <td>
          <p>
            Pending
          </p>
          <p />
        </td>
        <td>
          <ul>
            <li
              key="file-1"
            >
              File 1
            </li>
            <li
              key="file-2"
            >
              File 2
            </li>
          </ul>
        </td>
      </tr>
    </DataRequestTableRow>
  </tbody>
</table>
`;

exports[`DataRequestTableRow renders correctly 3`] = `
<table>
  <tbody>
    <DataRequestTableRow
      dataRequestPart={
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-1",
              "name": "File 1",
              "observation": Object {
                "id": "obs-1",
                "name": "Observation 1",
              },
            },
            Object {
              "id": "file-2",
              "name": "File 2",
              "observation": Object {
                "id": "obs-2",
                "name": "Observation 2",
              },
            },
          ],
          "id": "part-1",
          "status": "FAILED",
          "uri": "http://demo/data-request/part-1",
        }
      }
    >
      <tr>
        <td>
          Observation 1
        </td>
        <td>
          <p>
            Failed
          </p>
          <p>
            <button
              className="button re-request is-small is-danger is-rounded"
            >
              Re-request
            </button>
          </p>
        </td>
        <td>
          <ul>
            <li
              key="file-1"
            >
              File 1
            </li>
            <li
              key="file-2"
            >
              File 2
            </li>
          </ul>
        </td>
      </tr>
    </DataRequestTableRow>
  </tbody>
</table>
`;

exports[`DataRequestTableRow renders correctly 4`] = `
<table>
  <tbody>
    <DataRequestTableRow
      dataRequestPart={
        Object {
          "dataFiles": Array [
            Object {
              "id": "file-1",
              "name": "File 1",
              "observation": Object {
                "id": "obs-1",
                "name": "Observation 1",
              },
            },
            Object {
              "id": "file-2",
              "name": "File 2",
              "observation": Object {
                "id": "obs-2",
                "name": "Observation 2",
              },
            },
          ],
          "id": "part-1",
          "status": "EXPIRED",
          "uri": "http://demo/data-request/part-1",
        }
      }
    >
      <tr>
        <td>
          Observation 1
        </td>
        <td>
          <p>
            Expired
          </p>
          <p>
            <button
              className="button re-request is-small is-danger is-rounded"
            >
              Re-request
            </button>
          </p>
        </td>
        <td>
          <ul>
            <li
              key="file-1"
            >
              File 1
            </li>
            <li
              key="file-2"
            >
              File 2
            </li>
          </ul>
        </td>
      </tr>
    </DataRequestTableRow>
  </tbody>
</table>
`;
